File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Robot/ShooterSystem.java
package org.firstinspires.ftc.teamcode.Robot;

import org.firstinspires.ftc.teamcode.Robot.ShooterSubsystems.FlywheelSubsystem;
import org.firstinspires.ftc.teamcode.Robot.ShooterSubsystems.HoodSubsystem;
import org.firstinspires.ftc.teamcode.Robot.ShooterSubsystems.TurretSubsystem;

import dev.nextftc.core.subsystems.SubsystemGroup;
import dev.nextftc.ftc.ActiveOpMode;

public class ShooterSystem extends SubsystemGroup {
    public static ShooterSystem INSTANCE = new ShooterSystem();
    public ShooterSystem() {
        super(FlywheelSubsystem.INSTANCE, TurretSubsystem.INSTANCE, HoodSubsystem.INSTANCE);
    }

    public double[] getAngleVelocity(double distance) {
        if (distance < 2) {
            return new double[] {0,0.3};
        } else if (distance >= 2 && distance <3) {
            return new double[] {0,0.4};
        } else if (distance >= 3 && distance <4) {
            return new double[] {0,0.45};
        } else if (distance >= 4 && distance <5) {
            return new double[] {0.3,0.429};
        } else if (distance >= 5 && distance <6) {
            return new double[] {0.3,0.44};
        } else if (distance >= 6 && distance <7.5) {
            return new double[] {0.3,0.49};
        } else if (distance >= 7.5 && distance <8.5) {
            return new double[] {0.3,0.51};
        } else if (distance >= 8.5 && distance <9.5) {
            return new double[] {0.3,0.52};
        } else if (distance >= 9.5 && distance <10.5) {
            return new double[] {1,0.54};
        } else if (distance >= 10.5 && distance <11.5) {
            return new double[] {1,0.565};
        } else if (distance >= 11.5 && distance <12.5) {
            return new double[] {1,0.58};
        } else if (distance >= 12.5 && distance <13.5D) {
            return new double[] {1,0.614};
        } else {
            return new double[] {1,1650};
        }
    }

    public void autoAim(double x, double y, double heading) {
        double dx = 72-x;
        double dy = -72-y;
        double distance = Math.sqrt(x*x + y*y) / 12;
        double[] angleVel = getAngleVelocity(distance);
        TurretSubsystem.INSTANCE.setTurretHeading(dx, dy, heading);
        FlywheelSubsystem.INSTANCE.autoFlywheel(angleVel[1]);
        ActiveOpMode.telemetry().addData("Power", angleVel[1]);
        HoodSubsystem.INSTANCE.setPosition(angleVel[0]);
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Finals/Systems/ShooterSystem.kt
package org.firstinspires.ftc.teamcode.Finals.Systems

import com.pedropathing.geometry.Pose
import dev.nextftc.control.KineticState
import dev.nextftc.core.commands.Command
import dev.nextftc.core.commands.utility.InstantCommand
import dev.nextftc.core.subsystems.SubsystemGroup
import dev.nextftc.extensions.pedro.PedroComponent.Companion.follower
import dev.nextftc.ftc.ActiveOpMode
import org.firstinspires.ftc.teamcode.Finals.Systems.ShooterSystems.FlywheelSubsystem
import org.firstinspires.ftc.teamcode.Finals.Systems.ShooterSystems.HoodSubsystem
import org.firstinspires.ftc.teamcode.Finals.Systems.ShooterSystems.TurretSubsystem
import org.firstinspires.ftc.teamcode.Finals.Util.Alliance
import kotlin.math.PI
import kotlin.math.atan2
import kotlin.math.round
import kotlin.math.sqrt
import kotlin.properties.Delegates

object ShooterSystem: SubsystemGroup(FlywheelSubsystem, HoodSubsystem, TurretSubsystem) {
    val autoAimOnCommand: Command = InstantCommand { this.autoAimOn() }
    val autoAimOffCommand: Command = InstantCommand { this.autoAimOff() }

    const val flywheelEquationA: Double = -24.993970450383305
    const val flywheelEquationB: Double = -0.07154455960209152
    const val flywheelEquationC: Double = 0.00019278705563574583
    const val flywheelEquationD: Double = 116.26277827556967
    const val flywheelEquationE: Double = -0.272675259634784
    const val flywheelEquationF: Double = 122.99485320234648
    const val turretTicksPerRev: Double = (38450.0/24.0)
    lateinit var goalPose: Pose
    var turretLimit: Double by Delegates.notNull()

    var fullAutoAim: Boolean = false
    var partialAutoAimClose: Boolean = false
    var partialAutoAimFar: Boolean = false

    var shortVelocity: Double = 0.0;
    var farVelocity = 0.0;

    fun calibrateFlywheel(currPose: Pose) {
        val distanceFromGoal: Double = currPose.distanceFrom(goalPose)
        val hoodPosition: Double = HoodSubsystem.hoodServo.position
        val quadCoeffA: Double = flywheelEquationC
        val quadCoeffB: Double = (flywheelEquationB * hoodPosition) + flywheelEquationE
        val quadCoeffC: Double = (flywheelEquationA * hoodPosition * hoodPosition) + (flywheelEquationD * hoodPosition) + (flywheelEquationF - distanceFromGoal)
        val discriminant: Double = (quadCoeffB * quadCoeffB) - (4 * quadCoeffA * quadCoeffC)
        val velocity = (-1 * quadCoeffB + sqrt(discriminant)) / (2 * quadCoeffA)
        FlywheelSubsystem.flywheelControl.goal = KineticState(0.0,velocity)
    }
    fun calibrateTurret(currPose: Pose) {
        val angle = atan2(goalPose.x-currPose.x,goalPose.y-currPose.y)
        var ticks = (((currPose.heading-(PI/2))+angle) / (2*PI)) * turretTicksPerRev * -1
        ticks = normalizeTicks(ticks, 180.0)
        TurretSubsystem.turretControl.goal = KineticState(round(ticks))
        ActiveOpMode.telemetry.addData("Turret", "rawAngle = $angle")
        ActiveOpMode.telemetry.addData("Turret", "ticks = $ticks")
    }
    fun normalizeTicks(ticks: Double, angle: Double): Double {
        var ticks = ticks
        if (ticks < 0) {
            ticks = ticks + (turretTicksPerRev)
        }
        ticks %= (turretTicksPerRev)
        if (ticks > (turretTicksPerRev) * (angle / 360.0)) {
            ticks -= (turretTicksPerRev)
        }
        return ticks
    }

    fun calibrateHood(currPose: Pose) {
        val distanceFromGoal: Double = currPose.distanceFrom(goalPose)
        when (distanceFromGoal) {
            in 0.0..< 63.0 -> HoodSubsystem.hoodServo.position = 0.0
            in 63.0..<  105.0 -> HoodSubsystem.hoodServo.position = 0.65
            else              -> HoodSubsystem.hoodServo.position = 1.0
        }
    }

    fun autoAimOn() {
        fullAutoAim = true
        FlywheelSubsystem.flywheelAutoAim = true
    }
    fun autoAimOff() {
        fullAutoAim = false
        FlywheelSubsystem.flywheelAutoAim = false
    }

    fun setAlliance(alliance: Alliance) {
        this.goalPose = alliance.goalPose
    }

    override fun periodic() {
        if (fullAutoAim) {
            calibrateHood(follower.pose)
            calibrateTurret(follower.pose)
            calibrateFlywheel(follower.pose)
        } else if (partialAutoAimClose) {
            calibrateHood(follower.pose)
            calibrateTurret(follower.pose)
            FlywheelSubsystem.flywheelControl.goal = KineticState(0.0,shortVelocity)
        } else if (partialAutoAimFar) {
            calibrateHood(follower.pose)
            calibrateTurret(follower.pose)
            FlywheelSubsystem.flywheelControl.goal = KineticState(0.0,farVelocity)
        }
        ActiveOpMode.telemetry.addData("auto", this.fullAutoAim)
    }
}
